### PR TITLE
Report conflicting table as arg in transaction conflict exceptions

### PR DESCRIPTION
**Pull request by pkoenig10**
When a transaction fails due to a conflict and is retried, the logs do not indicate the table which had the conflict. This makes it difficult to diagnose the cause of excessive conflicts when retries ultimately allow the transaction to succeed.

### CODE REVIEW

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionConflictException.java [hunk 1]
```diff
@@ -81,6 +86,29 @@ public final String toString() {
...
+    private TransactionConflictException(
+            String message,
+            Collection<CellConflict> spanningWrites,
+            Collection<CellConflict> dominatingWrites,
+            TableReference conflictingTable) {
+        super(message);
+        this.spanningWrites = ImmutableList.copyOf(spanningWrites);
+        this.dominatingWrites = ImmutableList.copyOf(dominatingWrites);
+        this.conflictingTable = conflictingTable;
+        this.args = List.of(SafeArg.of("conflictingTable", conflictingTable));
```
Comment (ergo14) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionConflictException.java]: table references are not always safe, `LoggingArgs::tableRef`? 

From what I can tell the existing version funnels up the message (which seems / should be unsafe) to a normal runtime exception which should have its message logged as unsafe

Comment (pkoenig10) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionConflictException.java]: Yeah exception messages are always considered unsafe.

File: .palantir/revapi.yml [hunk 2]
```diff
@@ -87,6 +87,37 @@ acceptedBreaks:
       new: "parameter void com.palantir.atlasdb.keyvalue.api.RetryLimitReachedException::<init>(java.util.List<java.lang.Exception>,\
         \ ===java.util.List<com.palantir.atlasdb.keyvalue.api.RetryLimitReachedException.AttemptedTarget>===)"
       justification: "Only added recently"
+  "0.1079.0":
+    com.palantir.atlasdb:atlasdb-api:
+    - code: "java.method.numberOfParametersChanged"
```
Comment (pkoenig10) [inline .palantir/revapi.yml:100]: I only see 4 internal uses of these methods, all in tests.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 3]
```diff
@@ -2209,7 +2209,8 @@ protected void throwIfWriteAlreadyCommitted(
                         getStartTimestamp(),
                         spanningWrites,
                         dominatingWrites,
-                        System.currentTimeMillis() - timeCreated);
+                        System.currentTimeMillis() - timeCreated,
+                        List.of(LoggingArgs.tableRef("tableRef", tableRef)));
```
Comment (ergo14) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: The overload (`LoggingArgs.tableRef(tableRef)`) does the same thing?

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java [hunk 4]
```diff
@@ -16,25 +16,43 @@
 package com.palantir.atlasdb.transaction.api;
 
 import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.Safe;
+import com.palantir.logsafe.SafeLoggable;
+import java.util.List;
 
-public class TransactionSerializableConflictException extends TransactionFailedRetriableException {
+public class TransactionSerializableConflictException extends TransactionFailedRetriableException
```
Comment (ergo14) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java]: drive-by: should we make this final?

File: .palantir/revapi.yml (hunk 2 above)
Comment (ergo14) [reply .palantir/revapi.yml:100]: internal as in in atlas only?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadValidationCommitTimestampLoader.java [hunk 5]
```diff
@@ -124,7 +126,8 @@ private ListenableFuture<LongLongMap> getCommitTimestampsForTransactionsStartedA
                             "An uncommitted conflicting read was written after our start timestamp for table "
                                     + tableRef + ".  This case can cause deadlock and is very likely to be a "
                                     + "read write conflict.",
-                            tableRef);
+                            tableRef,
```
Comment (ergo14) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadValidationCommitTimestampLoader.java:129]: So tableRef is enough to construct the args. I am assuming we still pass the arg because adding the dep for LoggingArgs is awkward?

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java [hunk 6]
```diff
@@ -16,25 +16,43 @@
...
+    public TransactionSerializableConflictException(
+            String message, TableReference conflictingTable, List<Arg<?>> args) {
         super(message);
         this.conflictingTable = conflictingTable;
+        this.args = List.copyOf(args);
     }
 
     public static TransactionSerializableConflictException create(
-            TableReference tableRef, long timestamp, long elapsedMillis) {
+            TableReference tableRef, long timestamp, long elapsedMillis, List<Arg<?>> args) {
```
Comment (ergo14) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java:39]: We always pass only a list of one. I am assuming we use a list to support future cases where we want to add more args? Also, might be interesting to play with a variadic number of arguments for Arg instead of list - may make the test fixes look better

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java [hunk 7]
```diff
@@ -16,25 +16,43 @@
...
                         + " and another transaction wrote a different value than this transaction read.  startTs: %d "
                         + " elapsedMillis: %d",
                 tableRef.getQualifiedName(), timestamp, elapsedMillis);
-        return new TransactionSerializableConflictException(msg, tableRef);
+        return new TransactionSerializableConflictException(msg, tableRef, args);
+    }
+
+    @Override
+    public @Safe String getLogMessage() {
+        return "Transaction serializable conflict";
```
Comment (ergo14) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java]: Let's include more info here (something a bit more detailed, like the safe parts of the unsafe message)

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionConflictException.java [hunk 8]
```diff
@@ -33,7 +37,7 @@
  * The error message should be detailed about what caused the failure and what other transaction
  * conflicted with this one.
  */
-public final class TransactionConflictException extends TransactionFailedRetriableException {
+public final class TransactionConflictException extends TransactionFailedRetriableException implements SafeLoggable {
```
Comment (ergo14) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionConflictException.java:40]: most comments on the other exception class apply here

File: .palantir/revapi.yml (hunk 2 above)
Comment (pkoenig10) [reply .palantir/revapi.yml:100]: Internal as in, in non-public Palantir repos

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 3 above)
Comment (pkoenig10) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java]: Ah sorry, missed this one.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadValidationCommitTimestampLoader.java (hunk 5 above)
Comment (pkoenig10) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadValidationCommitTimestampLoader.java:129]: Couple reasons:
- `atlasdb-api` should not have a dependency on `atlasdb-client`
- We should give callers the flexibility to provide whatever args they want

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java (hunk 6 above)
Comment (pkoenig10) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java:39]: > Also, might be interesting to play with a variadic number of arguments for Arg instead of list - may make the test fixes look better

Up to you, but this is marginal more forwards compatible, since you can only have a single varargs method.

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java (hunk 7 above)
Comment (pkoenig10) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/TransactionSerializableConflictException.java:50]: Log messages should generally be compile time constants. Any additional information should be provided in args.

**Review by ergo14 — APPROVED** [review]
